File: server/datasets/java_source-code/Multiple-A3/control_flow_swapped_1.java
// Linked List Reversal - DISGUISE: control-flow-swapped
// Derived from organic_1.py. The iterative while-loop reversal has been
// rewritten as recursion.

class Node {
    int data;
    Node next;

    public Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedList {
    Node head;

    public LinkedList() {
        this.head = null;
    }

    public void append(int data) {
        Node newNode = new Node(data);
        if (this.head == null) {
            this.head = newNode;
            return;
        }
        Node current = this.head;
        while (current.next != null) {
            current = current.next;
        }
        current.next = newNode;
    }

    private Node reverseRecursive(Node curr, Node prev) {
        if (curr == null) {
            return prev;
        }
        Node nextNode = curr.next;
        curr.next = prev;
        return reverseRecursive(nextNode, curr);
    }

    public void reverse() {
        this.head = reverseRecursive(this.head, null);
    }

    public void display() {
        Node current = this.head;
        StringBuilder sb = new StringBuilder();
        while (current != null) {
            sb.append(current.data);
            if (current.next != null) {
                sb.append(" -> ");
            }
            current = current.next;
        }
        System.out.println(sb.toString());
    }

    public static void main(String[] args) {
        LinkedList ll = new LinkedList();
        for (int value : new int[]{1, 2, 3, 4, 5}) {
            ll.append(value);
        }
        System.out.println("Original list:");
        ll.display();
        ll.reverse();
        System.out.println("Reversed list:");
        ll.display();
    }
}

File: server/datasets/controlled_benchmarks/python/multiple_files/multiple_a3/control_flow_swapped_1.py
# Linked List Reversal - DISGUISE: control-flow-swapped
# Derived from organic_1.py. The iterative while-loop reversal has been
# rewritten as recursion, but it is the same algorithm producing the same
# output.

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None

    def append(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node

    def reverse(self):
        self.head = self._reverse_recursive(self.head, None)

    def _reverse_recursive(self, curr, prev):
        if curr is None:
            return prev
        next_node = curr.next
        curr.next = prev
        return self._reverse_recursive(next_node, curr)

    def display(self):
        values = []
        current = self.head
        while current:
            values.append(str(current.data))
            current = current.next
        print(" -> ".join(values))


if __name__ == "__main__":
    ll = LinkedList()
    for value in [1, 2, 3, 4, 5]:
        ll.append(value)

    print("Original list:")
    ll.display()

    ll.reverse()

    print("Reversed list:")
    ll.display()
